### PR TITLE
Recover parentheses in range patterns

### PR DESCRIPTION
**Pull request by ShE3py**
Before:
```rs
match n {
    (0).. => (),
    _ => ()
}
```
```
error: expected one of `=>`, `if`, or `|`, found `..`
 --> src/lib.rs:3:12
  |
3 |         (0).. => (),
  |            ^^ expected one of `=>`, `if`, or `|`
  ```
 After:
```
error: range pattern bounds cannot have parentheses
 --> main.rs:3:5
  |
3 |     (0).. => (),
  |     ^ ^
  |
help: remove these parentheses
  |
3 -     (0).. => (),
3 +     0.. => (),
  |
```

This sets the groundwork for #118625, which will extend the recovery to expressions like `(0 + 1)..` where users may tend to add parentheses to avoid dealing with precedence.

---

@rustbot label +A-parser +A-patterns +A-diagnostics

### CODE REVIEW

**Review by fmease — COMMENTED** [review]
I'm pretty sure this needs `.may_recover()` checks in some places not to regress code like:

```rs
macro_rules! m {
    ($pat:pat) => {};
    (($l:literal)..) => {};
}

fn main() {
    m!((0)..);
}
```

The code above now fails to compile or am I mistaken?

File: compiler/rustc_parse/src/parser/pat.rs [hunk 1]
```diff
@@ -727,6 +752,14 @@ impl<'a> Parser<'a> {
         begin: P<Expr>,
         re: Spanned<RangeEnd>,
     ) -> PResult<'a, PatKind> {
+        // recover from `(`
+        let open_paren = (self.may_recover()
+            && self.token.kind == token::OpenDelim(Delimiter::Parenthesis))
+        .then(|| {
+            self.bump();
+            self.prev_token.span
+        });
```
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:761]:
```suggestion
        let open_paren = (self.may_recover()
            && self.eat_noexpect(&token::OpenDelim(Delimiter::Parenthesis)))
        .then_some(self.prev_token.span);
```

File: compiler/rustc_parse/src/parser/pat.rs [hunk 2]
```diff
@@ -738,6 +771,19 @@ impl<'a> Parser<'a> {
             }
             None
         };
+
+        if let Some(span) = open_paren {
```
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:775]:
```suggestion
        if let Some(open_paren) = open_paren {
```

or `open_paren_sp` or sth. like that (as long as names for the opening and the closing parens are consistent >.<).

File: compiler/rustc_parse/src/errors.rs [hunk 3]
```diff
@@ -2378,6 +2378,27 @@ pub(crate) struct ExpectedCommaAfterPatternField {
     pub span: Span,
 }
 
+#[derive(Diagnostic)]
+#[diag(parse_unexpected_paren_in_range_pat)]
+pub(crate) struct UnexpectedParenInRangePat {
+    #[primary_span]
+    pub span: MultiSpan,
```
Comment (fmease) [inline compiler/rustc_parse/src/errors.rs:2385]:
```suggestion
    pub span: Vec<Span>,
```

File: compiler/rustc_parse/src/parser/pat.rs [hunk 4]
```diff
@@ -591,7 +594,29 @@ impl<'a> Parser<'a> {
...
+            let close_paren = self.prev_token.span;
+
+            match &pat.kind {
+                // recover ranges with parentheses around the `(start)..`
+                PatKind::Lit(begin)
+                    if self.may_recover()
+                        && let Some(form) = self.parse_range_end() =>
+                {
+                    self.dcx().emit_err(UnexpectedParenInRangePat {
+                        span: MultiSpan::from_spans(vec![open_paren, close_paren]),
```
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:607]:
```suggestion
                        span: vec![open_paren, close_paren],
```

together with the `MultiSpan`→`Vec<Span>` suggestion.

File: compiler/rustc_parse/src/parser/pat.rs [hunk 5]
```diff
@@ -738,6 +771,19 @@ impl<'a> Parser<'a> {
             }
             None
         };
+
+        if let Some(span) = open_paren {
+            self.expect(&token::CloseDelim(Delimiter::Parenthesis))?;
+
+            self.dcx().emit_err(UnexpectedParenInRangePat {
+                span: MultiSpan::from_spans(vec![span, self.prev_token.span]),
```
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:779]:
```suggestion
                span: vec![span, self.prev_token.span],
```

ditto

File: compiler/rustc_parse/src/parser/pat.rs [hunk 6]
```diff
@@ -777,11 +823,32 @@ impl<'a> Parser<'a> {
...
+            self.prev_token.span
+        });
+
         let end = self.parse_pat_range_end()?;
         if let RangeEnd::Included(syn @ RangeSyntax::DotDotDot) = &mut re.node {
             *syn = RangeSyntax::DotDotEq;
             self.dcx().emit_err(DotDotDotRangeToPatternNotAllowed { span: re.span });
         }
+
+        if let Some(span) = open_paren {
```
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:840]:
```suggestion
        if let Some(open_span) = open_paren {
```

ditto

File: compiler/rustc_parse/src/parser/pat.rs [hunk 7]
```diff
@@ -777,11 +823,32 @@ impl<'a> Parser<'a> {
     /// The form `...X` is prohibited to reduce confusion with the potential
     /// expression syntax `...expr` for splatting in expressions.
     fn parse_pat_range_to(&mut self, mut re: Spanned<RangeEnd>) -> PResult<'a, PatKind> {
+        // recover from `(`
+        let open_paren = (self.may_recover()
+            && self.token.kind == token::OpenDelim(Delimiter::Parenthesis))
+        .then(|| {
+            self.bump();
+            self.prev_token.span
+        });
```
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:832]:
```suggestion
        let open_paren = (self.may_recover()
            && self.eat_noexpect(&token::OpenDelim(Delimiter::Parenthesis)))
        .then_some(self.prev_token.span);
```

File: compiler/rustc_parse/src/parser/pat.rs [hunk 8]
```diff
@@ -942,7 +1013,7 @@ impl<'a> Parser<'a> {
 
         if self.isnt_pattern_start() {
             let descr = super::token_descr(&self.token);
-            self.dcx().emit_err(errors::BoxNotPat {
+            self.dcx().emit_err(BoxNotPat {
```
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:1016]: Could you please undo this change and un-import the struct?

File: tests/ui/half-open-range-patterns/range_pat_interactions2.rs [hunk 9]
```diff
@@ -8,8 +8,7 @@ fn main() {
     for x in -9 + 1..=(9 - 2) {
         match x as i32 {
             0..=(5+1) => errors_only.push(x),
-            //~^ error: inclusive range with no end
-            //~| error: expected one of `=>`, `if`, or `|`, found `(`
+            //~^ error: expected `)`, found `+`
```
Comment (fmease) [inline tests/ui/half-open-range-patterns/range_pat_interactions2.rs:11]: Ah, this is not ideal but you probably plan on fixing this in a follow-up PR or in #118625 (I haven't looked at the changes in that PR)?

We could create parser snapshots and try to parse a *RangePatternBound* first, then a *Pattern* on failure and finally an *Expression* on failure and provide targeted diagnostics accordingly.

**Review by fmease — CHANGES_REQUESTED** [review]

**Comment by fmease** [comment]
r? fmease

File: tests/ui/half-open-range-patterns/range_pat_interactions2.rs (hunk 9 above)
Comment (ShE3py) [reply tests/ui/half-open-range-patterns/range_pat_interactions2.rs]: I changed it to say « unexpected `(` » if *RangePatternBound* didn't parse or if it did but didn't end with `)` , but yeah #118625 will try to parse an *Expression* to emit a better diagnostic:
```
error: expected pattern bound, found expression
  --> $DIR/range_pat_interactions2.rs:10:18
   |
LL |             0..=(5+1) => errors_only.push(x),
   |                  ^^^ expressions are not allowed in patterns

error: range pattern bounds cannot have parentheses
  --> $DIR/range_pat_interactions2.rs:10:17
   |
LL |             0..=(5+1) => errors_only.push(x),
   |                 ^   ^
   |
help: remove these parentheses
   |
LL -             0..=(5+1) => errors_only.push(x),
LL +             0..=5+1 => errors_only.push(x),
   |
   ```
However #118625 got a bit too big, so I split it into sub-PRs and I'll rebase it once this one has been merged.

**Review by fmease — APPROVED** [review]
Thanks for applying my suggestions!

I've had a think about https://github.com/rust-lang/rust/pull/119397#discussion_r1438450405 again and while that diagnostic (`` expected `)`, found `+` ``) isn't ideal, you plan on fixing that soon anyway . Now, the latest version of your PR contains two ad hoc diagnostics (``unexpected token: `(` ``) which aren't ideal either.

So let's go back to bailing out early with `?` if we fail to parse `bound` in `parse_pat_range_end` and to `expect`'ing `)` instead of `eat_noexpect`'ing to eliminate both occurrences of `struct_span_err`.

After, that this PR is good to go! Sorry for taking a while to respond. I'm excited to see you improving the parse errors for patterns! :heart:
